internal/expr/query.go

```diff
@@ -132,19 +160,45 @@ func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error)
 			continue
 		}
 		if idx >= len(qe.outputs) {
-			return nil, fmt.Errorf("internal error: sqlair column not in outputs (%d>=%d)", idx, len(qe.outputs))
+			return nil, nil, fmt.Errorf("internal error: sqlair column not in outputs (%d>=%d)", idx, len(qe.outputs))
 		}
-		field := qe.outputs[idx]
-		outputVal, ok := typeDest[field.structType]
-		if !ok {
-			return nil, fmt.Errorf("type %q found in query but not passed to decode", field.structType.Name())
+		typeElement := qe.outputs[idx]
+		switch te := typeElement.(type) {
+		case field:
+			outputVal, ok := typeDest[te.structType]
+			if !ok {
+				return nil, nil, fmt.Errorf("type %q found in query but not passed to decode", te.structType.Name())
+			}
+			val := outputVal.FieldByIndex(te.index)
+			if !val.CanSet() {
+				return nil, nil, fmt.Errorf("internal error: cannot set field %s of struct %s", te.name, te.structType.Name())
+			}
+			ptrs = append(ptrs, val.Addr().Interface())
+		case mapKey:
+			mapDecodeInfo, ok := mapDecodeInfos[te.mapType]
+			if !ok {
+				return nil, nil, fmt.Errorf("type %q found in query but not passed to decode", te.mapType.Name())
+			}
+			// Scan in to a new variable with the type of the maps value
+			val := reflect.New(te.mapType.Elem()).Elem()
+			addr := val.Addr().Interface()
+			ptrs = append(ptrs, addr)
+			mapDecodeInfo.valuePtrs = append(mapDecodeInfo.valuePtrs, addr)
+			mapDecodeInfo.keyIndex[te.name] = len(mapDecodeInfo.valuePtrs) - 1
 		}
+	}
+	var mapInfoSlice = []*MapDecodeInfo{}
+	for _, v := range mapDecodeInfos {
+		mapInfoSlice = append(mapInfoSlice, v)
+	}
+	return ptrs, mapInfoSlice, nil
+}
 
-		val := outputVal.FieldByIndex(field.index)
-		if !val.CanSet() {
-			return nil, fmt.Errorf("internal error: cannot set field %s of struct %s", field.name, field.structType.Name())
-		}
-		ptrs = append(ptrs, val.Addr().Interface())
+// Populate puts scanned values into the map.
+func (mapDecodeInfo *MapDecodeInfo) Populate() {
+	for k, i := range mapDecodeInfo.keyIndex {
+		// Is there a better way to pass the set values here than their
+		// pointers?
+		mapDecodeInfo.m.SetMapIndex(reflect.ValueOf(k), reflect.ValueOf(mapDecodeInfo.valuePtrs[i]).Elem())
```

Comment:
As in the previous comment, maybe we could instead do (please see previous comment for `kv` here):

```Go
mapDecodeInfo.m.SetMapIndex(reflect.ValueOf(k), reflect.ValueOf(mapDecodeInfo.kv[k])
```

Reply:
Changed!